nrf52-hal-common/src/twim.rs

```diff
@@ -0,0 +1,258 @@
+//! HAL interface to the SPIM peripheral
+//!
+//! See product specification:
+//!
+//! - nrf52832: Section 33
+//! - nrf52840: Section 6.31
+
+
+use core::ops::Deref;
+
+use target::{
+    twim0,
+    P0,
+    TWIM0,
+    TWIM1,
+};
+
+use gpio::{
+    p0::P0_Pin,
+    Floating,
+    Input,
+};
+
+
+pub use target::twim0::frequency::FREQUENCYW as Frequency;
+
+
+pub trait TwimExt: Deref<Target=twim0::RegisterBlock> + Sized {
+    fn constrain(self, pins: Pins, frequency: Frequency)
+        -> Twim<Self>;
+}
+
+macro_rules! impl_twim_ext {
+    ($($twim:ty,)*) => {
+        $(
+            impl TwimExt for $twim {
+                fn constrain(self, pins: Pins, frequency: Frequency)
+                    -> Twim<Self>
+                {
+                    Twim::new(self, pins, frequency)
+                }
+            }
+        )*
+    }
+}
+
+impl_twim_ext!(
+    TWIM0,
+    TWIM1,
+);
+
+
+/// Interface to a TWIM instance
+///
+/// This is a very basic interface that comes with the following limitation:
+/// The TWIM instances share the same address space with instances of SPIM,
```

Comment:
It might be worth clarifying that the address space is shared with SPIMn, TWISn, etc.

For example, we can use SPIM0 with TWIM1, but not SPIM1 with TWIM1.

In the future, it might make sense to have these peripheral groups in a wrapped group in the HAL Peripherals block, so you can only take one at a time, e.g. something like:

```rust
struct PeripheralGroup3 {
    SPIM0: ...,
    SPIS0: ...,
    TWIM0: ...,
    ...,
}

impl PeripheralGroup3 {
pub fn into_spim_0(self) -> SPIM0 {}
// ...
}

Reply:
I agree. Also see https://github.com/nrf-rs/nrf52-hal/issues/7, as that related to solving this issue and similar ones.

I'll update the documentation. I need to update the commit anyway, as I just noticed I wrote "SPIM" in the message where I should have written "TWIM".